### ISSUE TITLE
Custom Key and Trust Stores

### ISSUE THREAD

**Issue by jonathanamartin** [cloudfoundry/java-buildpack#378]
I have a need to provide dev teams with the ability to provide client and server keys and certificates to their applications running in Cloud Foundry.

We've used the overlay method to install JCE unlimited strength policy files and corporate certificates.  However, for mutual authentication the key and trust stores will change for every app.

We're in a controlled environment so developers can't use SSH in to drop files.  However, they can pass app environment variables.

I've looked at the [cache extension](https://github.com/cloudfoundry/java-buildpack/blob/master/docs/extending-caches.md).  Also thought about [providing a JAR](https://github.com/cloudfoundry/java-buildpack/issues/27) to let the application dynamically update the certificate stores.

However, I'm thinking we might be able to use the following approach since we've already got a fork of the Java Buildpack:

1. Dev teams provide PEM encoded trust and key stores as app environment variables.
2. Add an additional `ModularComponent` to the Tomcat Container module that:
    1. Converts the PEM stores to JKS.
    2. Either updates `server.xml` or passes the key and trust store as JVM options.

Does this sound like a viable approach or is there a more elegant way to do this?

**Update**
Minor clarification.  I've proven out mutual auth with SSL using an embedded Tomcat with key stores on the classpath.  However, due to our corporate patching policies we need to use the Tomcat container provided in the Java Buildpack.

**Comment by jonathanamartin** [comment]
Any thoughts on this?

**Comment by nebhale** [comment]
We generally recommend that users include a KeyStore/TrustStore in their application (typically in `META-INF` so that it can't be accidentally served as a resource) and then reference it via `$PWD/app/META-INF/keystore.jks`.  Is there a reason that this doesn't work for your applications?

**Comment by jonathanamartin** [comment]
That could be a possibility.  Though we've got a corporate policy to manage JKS files in a separate certificate management tool.  Also, our security folks aren't in favor of storing production keys with the app source.  Might be able to merge them or something at build time.

However, the goal is to use client certificate authentication over SSL/TSL managed by Tomcat with TCP routes.  Tomcat needs to be external so it can be patched on the corporate schedule.

At a minimum is there a way to configure the connector in `server.xml` enabling SSL and setting the port to `8080`?

**Comment by nebhale** [comment]
@jonathanamartin Any Tomcat configuration can be made by adding [a `server.xml` to the `resources/tomcat/conf` folder of a fork](https://github.com/cloudfoundry/java-buildpack/blob/master/docs/container-tomcat.md#additional-resources).  Since this seems to be the kind of thing that is broadly applicable across all of the applications in an enterprise, this seems to be the right situation for a fork.

**Comment by jonathanamartin** [comment]
@nebhale thanks for the update.  We've got a fork and have made modifications to `server.xml`.  However, for mutual authentication the key and trust stores will change for every app.

Since we've already got a fork I think I'll try the following:
1. Dev teams provide PEM encoded trust and key stores as app environment variables.
2. Add an additional `ModularComponent` to the Tomcat Container module that:
    1. Converts the PEM stores to JKS.
    2. Either updates `server.xml` or passes the key and trust store as JVM options.

If I can pull it off does this sound like a viable solution?  Would there be any downside to taking this approach?
